packages/router/src/grpc/custody/authorize/assert-valid-swaps.ts

```diff
@@ -0,0 +1,19 @@
+import { Address } from '@buf/penumbra-zone_penumbra.bufbuild_es/penumbra/core/keys/v1/keys_pb';
+import { TransactionPlan } from '@buf/penumbra-zone_penumbra.bufbuild_es/penumbra/core/transaction/v1/transaction_pb';
+import { Code, ConnectError } from '@connectrpc/connect';
+
+export const assertValidSwaps = (
+  plan: TransactionPlan,
+  isControlledAddress: (address?: Address) => boolean,
+): void => {
+  plan.actions.forEach(action => {
+    if (action.action.case !== 'swap') return;
+
+    if (!isControlledAddress(action.action.value.swapPlaintext?.claimAddress)) {
+      throw new ConnectError(
+        "Tried to initiate a swap with a claim address belonging to a different user. This means that, when the swap is claimed, the funds would go to someone else's address, not yours. This should never happen. The website you are using may be trying to steal your funds!",
+        Code.PermissionDenied,
+      );
```

Comment:
This seems like quite a helpful error message. Can we show this in our tooltip instead of it being swallowed?

Reply:
This is related to @TalDerei 's comment above:

> we might want to consider displaying a more stateful error message describing why the swap transaction was canceled down the line.

Here's the issue: minifront will never try to steal the user's funds, so we never need to show this message to minifront users.

We _should_ show this message to users on other sites who try to steal the user's funds, but obviously, a malicious website isn't going to proactively show this error to people. But if they're trying to steal people's funds, they'll presumably test the stealing-people's-funds functionality first, and they'll realize that this error shows up, at which point they'll either A) give up, or B) at least suppress this error.

So I'm not actually sure how useful this error is?

The only way it could be useful is if we could show this error _in the transaction approval popup_, which the malicious website has no control over. Not sure if this is possible, though, since this error is thrown inside an RPC method implementation. What if we merge this as-is for now, and then revisit this question in a separate ticket?

Reply:
created separate ticket to limit scope